<a href="https://colab.research.google.com/github/Keltings/ADS-Assignment-1/blob/main/Accident_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**1. DATA OVERVIEW

```
`# This is formatted as code`
```

**

In [1]:
#import packages for data processing
import pandas as pd
import numpy as np

In [2]:
#In the first place we are going to import the dataset using pandas.
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/accidents 1.csv')

In [3]:
df.shape #check the number of rows and columns the dataset has

(867550, 33)

In [14]:
df.info() #show entry, datatype memory usage

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 867550 entries, 0 to 867549
Data columns (total 33 columns):
 #   Column                                       Non-Null Count   Dtype  
---  ------                                       --------------   -----  
 0   Accident_Index                               867550 non-null  object 
 1   Location_Easting_OSGR                        867464 non-null  float64
 2   Location_Northing_OSGR                       867464 non-null  float64
 3   Longitude                                    867464 non-null  float64
 4   Latitude                                     867464 non-null  float64
 5   Police_Force                                 867550 non-null  int64  
 6   Accident_Severity                            867550 non-null  int64  
 7   Number_of_Vehicles                           867550 non-null  int64  
 8   Number_of_Casualties                         867550 non-null  int64  
 9   Date                                         867550 non-nul

The dataset has 867550 entries and 32 features + 1 target variable. Automatic Type Recognition: 13 of them are int type, 5 of them are float64, 15 of them are object. Memory usage: 218.4+ MB.



> **STATISTICS VIEW**



In [17]:
#Basic statistic on Ordinal, Interval and Ratio data.
OIR_columns = ['Location_Easting_OSGR', 'Location_Northing_OSGR', 'Longitude', 'Latitude', 'Police_Force', 'Accident_Severity', 'Number_of_Vehicles', 'Number_of_Casualties', 'Day_of_Week', 'Local_Authority_(District)', '1st_Road_Class', '1st_Road_Number', 'Speed_limit', '2nd_Road_Class', '2nd_Road_Number', 'Urban_or_Rural_Area', 'Year']
df[OIR_columns].describe() #Exclude Date; Time can be transformed to second/minute/hour/day/month/year separately, but it is no need to do it here.


Location_Easting_OSGR  ...           Year
count          867464.000000  ...  867550.000000
mean           439345.364558  ...    2007.933220
std             95453.947190  ...       2.326594
min             64950.000000  ...    2005.000000
25%            375220.000000  ...    2006.000000
50%            439790.000000  ...    2009.000000
75%            523520.000000  ...    2010.000000
max            655370.000000  ...    2011.000000

[8 rows x 17 columns]

What can be derived from the above is that 2nd_Road_Class: 25%,50% and 75%are all zero which is quite abnormal
Junction_Detail: has null everything so should not be used for our analysis

In [18]:
#Basic statistics on the rest of the data

df.loc[:, ~df.columns.isin(OIR_columns)].astype("object").describe() # All the Nominal data can be treated as "object" type for simplicity.

Accident_Index  ... LSOA_of_Accident_Location
count          867550  ...                    808937
unique         558085  ...                     34140
top     2010000000000  ...                 E01000004
freq           137503  ...                      1749

[4 rows x 16 columns]

Road_Type: Single carriageway seems to be a common road name that an accidents would happen. The name of the type of the road could mean something or contain some special characters.
Light_Conditions: It seems that accidents usually happen when on daylight when street lights are present.
Weather_Conditions: Fine without high winds  is the most frequent weather condition.

**Data Categorizing**

In [27]:
# Category Visualization
# https://plotly.com/python/sunburst-charts/
df_category = dict(
    character=['Basic', 'Location', 'Environment', 'Infrastructure', 'Accident_Index', 'Location_Easting_OSGR', 'Location_Northing_OSGR',
               'Longitude', 'Latitude', 'Police_Force', 'Accident_Severity',
               'Number_of_Vehicles', 'Number_of_Casualties', 'Date', 'Day_of_Week',
               'Time', 'Local_Authority_(District)', 'Local_Authority_(Highway)',
               '1st_Road_Class', '1st_Road_Number', 'Road_Type', 'Speed_limit',
               'Junction_Detail', 'Junction_Control', '2nd_Road_Class',
               '2nd_Road_Number', 'Pedestrian_Crossing-Human_Control',
               'Pedestrian_Crossing-Physical_Facilities', 'Light_Conditions',
               'Weather_Conditions', 'Road_Surface_Conditions',
               'Special_Conditions_at_Site', 'Carriageway_Hazards',
               'Urban_or_Rural_Area', 'Did_Police_Officer_Attend_Scene_of_Accident',
               'LSOA_of_Accident_Location', 'Year'],
     parent = ["","","","","Basic", "Location", "Location", "Location", "Location", "Basic", "Basic", "Basic",
               "Basic", "Basic", "Basic", "Basic", "Infrastructure", "Infrastructure", "Infrastructure", "Infrastructure", 
               "Infrastructure", "Infrastructure", "Infrastructure", "Infrastructure", "Infrastructure", "Infrastructure", 
               "Infrastructure", "Infrastructure", "Environment", "Environment", "Environment", "Environment", 
               "Environment", "Location", "Basic", "Basic", "Basic"],
)

category_fig =px.sunburst(
    df_category,
    names='character',
    parents='parent'
)

category_fig.update_layout(
    autosize=False,
    width=1050,
    height=550,
    title={
        'text': "Data Categorizing",
        'y':0.95,
        'x':0.5,
        'xanchor': 'center',
        'yanchor': 'top'})
category_fig.show()

**2. DATA PREPROCESSING**

Here we perform the art of cleaning our data set  for better descriptive and predictive analysis
 

In [ ]:
#h

In [19]:
df.columns

Index(['Accident_Index', 'Location_Easting_OSGR', 'Location_Northing_OSGR',
       'Longitude', 'Latitude', 'Police_Force', 'Accident_Severity',
       'Number_of_Vehicles', 'Number_of_Casualties', 'Date', 'Day_of_Week',
       'Time', 'Local_Authority_(District)', 'Local_Authority_(Highway)',
       '1st_Road_Class', '1st_Road_Number', 'Road_Type', 'Speed_limit',
       'Junction_Detail', 'Junction_Control', '2nd_Road_Class',
       '2nd_Road_Number', 'Pedestrian_Crossing-Human_Control',
       'Pedestrian_Crossing-Physical_Facilities', 'Light_Conditions',
       'Weather_Conditions', 'Road_Surface_Conditions',
       'Special_Conditions_at_Site', 'Carriageway_Hazards',
       'Urban_or_Rural_Area', 'Did_Police_Officer_Attend_Scene_of_Accident',
       'LSOA_of_Accident_Location', 'Year'],
      dtype='object')

In [20]:
#data visualization
import matplotlib.pyplot as plt
%matplotlib inline
from pandas_profiling import ProfileReport
import plotly.express as px
import plotly.figure_factory as ff
import datetime as dt
import plotly.graph_objects as go
from plotly.subplots import make_subplots

In [6]:
numerics = ['int64','float64']
numeric_df = df.select_dtypes(include=numerics)
len(numeric_df.columns)

18

The below codes were set up for pandas profiling which would ease our work of analysing the dataset

In [ ]:
#@title Default title text
data = df
data = ff.create_table(data.head(6))
data

In [ ]:
!pip uninstall pandas_profiling

In [ ]:
!pip install pandas-profiling[notebook,html]

In [ ]:
profile = ProfileReport(df, title="accident analysis")
profile

In [ ]:
profile.to_notebook_iframe()

In [ ]:
profile.to_file(output_file="accident analysis.html")

In [ ]:
profile.to_file(output_file="accident analysis.csv")

In [ ]:
profile.to_file(output_file="your_report.json")

Use of pandas profiling helps in analysing yur data more faster. Notice that we have missing values hence our data needs cleaning

In [16]:
#rename the columns for a clear dataset and easy reading
accident_data = df.rename(columns={'Accident_Index': 'Index', 'Location_Easting_OSGR': 'LEO', 'Location_Northing_OSGR': 'LNO',
       'Longitude': 'Long', 'Latitude': 'Lat', 'Police_Force': 'PF', 'Accident_Severity': 'AS',
       'Number_of_Vehicles': 'NV', 'Number_of_Casualties': 'NC', 'Date': 'Date', 'Day_of_Week': 'DW',
       'Time': 'Time', 'Local_Authority_(District)': 'District', 'Local_Authority_(Highway)': 'Highway',
       '1st_Road_Class': 'RC1', '1st_Road_Number': 'RN1', 'Road_Type': 'RT', 'Speed_limit': 'SL',
       'Junction_Detail': 'JD', 'Junction_Control': 'JC', '2nd_Road_Class': 'RC2',
       '2nd_Road_Number': 'RN2', 'Pedestrian_Crossing-Human_Control': 'PCHC',
       'Pedestrian_Crossing-Physical_Facilities': 'PCPF', 'Light_Conditions': 'LC',
       'Weather_Conditions': 'WC', 'Road_Surface_Conditions': 'RSC',
       'Special_Conditions_at_Site': 'SCS', 'Carriageway_Hazards': 'CH',
       'Urban_or_Rural_Area': 'URA', 'Did_Police_Officer_Attend_Scene_of_Accident': 'DPOASA',
       'LSOA_of_Accident_Location': 'LSOAAL', 'Year': 'Year'})
accident_data.head()

Index       LEO       LNO      Long  ...  URA  DPOASA     LSOAAL  Year
0  200901BS70001  524910.0  180800.0 -0.201349  ...    1     Yes  E01002882  2009
1  200901BS70002  525050.0  181040.0 -0.199248  ...    1     Yes  E01002886  2009
2  200901BS70003  526490.0  177990.0 -0.179599  ...    1     Yes  E01002912  2009
3  200901BS70004  524800.0  180300.0 -0.203110  ...    1     Yes  E01002871  2009
4  200901BS70005  526930.0  177490.0 -0.173445  ...    1     Yes  E01002840  2009

[5 rows x 33 columns]

In [ ]:
!pip install category_encoders

Required Libraries

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import category_encoders as ce

from sklearn.metrics import classification_report, confusion_matrix, accuracy_score, f1_score, roc_curve
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.neural_network import MLPClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier, plot_tree
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import precision_recall_curve
from sklearn.metrics import average_precision_score
from sklearn.preprocessing import label_binarize

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning:

pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.



In [ ]:
import plotly.graph_objects as go
from nltk.corpus import stopwords

Exploratory Data Analysis (EDA)
Exploratory Data Analysis refers to the critical process of performing initial investigations on data so as to discover patterns,to spot anomalies,to test hypothesis and to check assumptions with the help of summary statistics and graphical representations. In this first part of this notebook we are going to do an EDA of the dataset.

In [8]:
print(accident_data["AS"].value_counts())

3    741295
2    114777
1     11478
Name: AS, dtype: int64


In [9]:
accident_data.columns

Index(['Index', 'LEO', 'LNO', 'Long', 'Lat', 'PF', 'AS', 'NV', 'NC', 'Date',
       'DW', 'Time', 'District', 'Highway', 'RC1', 'RN1', 'RT', 'SL', 'JD',
       'JC', 'RC2', 'RN2', 'PCHC', 'PCPF', 'LC', 'WC', 'RSC', 'SCS', 'CH',
       'URA', 'DPOASA', 'LSOAAL', 'Year'],
      dtype='object')

In [10]:
accident_data1= accident_data
accident_data1.shape

(867550, 33)

In [11]:
accident_data1["Date"] = pd.to_datetime(accident_data1["Date"])

# Extract year, month, weekday and day
accident_data1["Year"] = accident_data1["Date"].dt.year
accident_data1["Month"] = accident_data1["Date"].dt.month
accident_data1["Weekday"] = accident_data1["Date"].dt.weekday
accident_data1["Day"] = accident_data1["Date"].dt.day

# Extract hour and minute
accident_data1['Time'] = pd.to_datetime(accident_data1['Time'])
accident_data1["Hour"] = accident_data1["Time"].dt.hour
accident_data1["Minute"] = accident_data1["Time"].dt.minute

accident_data.head(3)

Index       LEO       LNO      Long  ...  Weekday  Day  Hour  Minute
0  200901BS70001  524910.0  180800.0 -0.201349  ...        3    1  15.0    11.0
1  200901BS70002  525050.0  181040.0 -0.199248  ...        4    1  10.0    59.0
2  200901BS70003  526490.0  177990.0 -0.179599  ...        2    1  14.0    19.0

[3 rows x 38 columns]

In [12]:
accident_data1 = accident_data1.drop(['Date', 'Time', 'JD', 'Weekday'], axis = 1)
accident_data1.shape


(867550, 34)

In [13]:
# dropping duplicate values
accident_data1.drop_duplicates(keep=False,inplace=True)
accident_data1.shape

(867534, 34)